mempool/src/pool.rs

```diff
@@ -0,0 +1,1093 @@
+use common::chain::OutPoint;
+use common::primitives::amount::Amount;
+use common::primitives::Id;
+use common::primitives::Idable;
+use common::primitives::H256;
+
+use logging::log;
+
+use utils::newtype;
+
+use crate::error::Error;
+use crate::error::TxValidationError;
+use crate::feerate::FeeRate;
+use crate::feerate::INCREMENTAL_RELAY_FEE_RATE;
+use crate::feerate::INCREMENTAL_RELAY_THRESHOLD;
+
+const ROLLING_FEE_BASE_HALFLIFE: Time = Duration::new(60 * 60 * 12, 1);
+// TODO this willbe defined elsewhere (some of limits.rs file)
+const MAX_BLOCK_SIZE_BYTES: usize = 1_000_000;
+
+const MAX_BIP125_REPLACEMENT_CANDIDATES: usize = 100;
+
+// TODO this should really be taken from some global node settings
+const RELAY_FEE_PER_BYTE: usize = 1;
+
+const MAX_MEMPOOL_SIZE_BYTES: usize = 300_000_000;
+
+const DEFAULT_MEMPOOL_EXPIRY: Duration = Duration::new(336 * 60 * 60, 0);
+
+const ROLLING_FEE_DECAY_INTERVAL: Time = Duration::new(10, 0);
+
+pub(crate) type MemoryUsage = usize;
+
+#[automock]
+pub trait GetMemoryUsage: 'static {
+    fn get_memory_usage(&self) -> MemoryUsage;
+}
+
+pub(crate) type Time = Duration;
+pub trait GetTime: Clone + 'static {
+    fn get_time(&self) -> Time;
+}
+
+impl<C, T, M> TryGetFee for MempoolImpl<C, T, M>
+where
+    C: ChainState,
+    T: GetTime,
+    M: GetMemoryUsage,
+{
+    fn try_get_fee(&self, tx: &Transaction) -> Result<Amount, TxValidationError> {
+        let inputs = tx
+            .inputs()
+            .iter()
+            .map(|input| {
+                let outpoint = input.outpoint();
+                self.chain_state
+                    .get_outpoint_value(outpoint)
+                    .or_else(|_| self.store.get_unconfirmed_outpoint_value(outpoint))
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+        let sum_inputs = inputs
+            .iter()
+            .cloned()
+            .sum::<Option<_>>()
+            .ok_or(TxValidationError::InputValuesOverflow)?;
+        let sum_outputs = tx
+            .outputs()
+            .iter()
+            .map(|output| output.value())
+            .sum::<Option<_>>()
+            .ok_or(TxValidationError::OutputValuesOverflow)?;
+        (sum_inputs - sum_outputs).ok_or(TxValidationError::InputsBelowOutputs)
+    }
+}
+
+fn get_relay_fee(tx: &Transaction) -> Amount {
+    // TODO we should never reach the expect, but should this be an error anyway?
+    eprintln!("get_relay_fee - encoded_size {:?}", tx.encoded_size());
+    Amount::from_atoms(u128::try_from(tx.encoded_size() * RELAY_FEE_PER_BYTE).expect("Overflow"))
+}
+
+pub trait Mempool<C, T, M> {
+    fn create(chain_state: C, clock: T, memory_usage_estimator: M) -> Self;
+    fn add_transaction(&mut self, tx: Transaction) -> Result<(), Error>;
+    fn get_all(&self) -> Vec<&Transaction>;
+    fn contains_transaction(&self, tx: &Id<Transaction>) -> bool;
+    fn drop_transaction(&mut self, tx: &Id<Transaction>);
+    fn new_tip_set(&mut self, chain_state: C);
+}
```

Comment:
Added some basic docs. Let me know if there's anything you'd like to see added.

Reply:
Two more possible improvements:
* Use doc comments, starting with `///` rather than `//`
* The comment on `drop_transaction` would be better to have the first sentence in one paragraph and the rest in another paragraph. That's because the first part would be used by `cargo doc` as a brief description that shows up in the list of trait methods and the full description will be only displayed if you click on it for more details